dimod/include/dimod/abc.h

```diff
@@ -0,0 +1,1070 @@
+    iterator erase(iterator first, iterator last) { return neighborhood_.erase(first, last); }
+
+    /// Return an iterator to the first element that does not come before `v`.
+    iterator lower_bound(index_type v) {
+        return std::lower_bound(begin(), end(), v);
+    }
+
+    /// Return an iterator to the first element that does not come before `v`.
+    const_iterator lower_bound(index_type v) const {
+        return std::lower_bound(cbegin(), cend(), v);
+    }
+
+    /**
+     * Total bytes consumed by the biases and indices.
+     *
+     * If `capacity` is true, use the capacity of the underlying vectors rather
+     * than the size.
+     */
+    size_type nbytes(bool capacity = false) const noexcept {
+        // so there is no guaruntee that the compiler will not implement
+        // pair as pointers or whatever, but this seems like a reasonable
+        // assumption.
+        if (capacity) {
+            return neighborhood_.capacity() * sizeof(value_type);
+        } else {
+            return neighborhood_.size() * sizeof(value_type);
+        }
+    }
+
+    /**
+     * Return the bias at neighbor `v` or the default value.
+     *
+     * Return the bias of `v` if `v` is in the neighborhood, otherwise return
+     * the `value` provided without inserting `v`.
+     */
+    bias_type get(index_type v, bias_type value = 0) const {
+        auto it = lower_bound(v);
+
+        if (it != cend() && it->v == v) {
+            // it exists
+            return it->bias;
+        } else {
+            // it doesn't exist
+            return value;
+        }
+    }
+
+    /// Request that the neighborhood capacity be at least enough to contain `n`
+    /// elements.
+    void reserve(index_type n) { neighborhood_.reserve(n); }
+
+    /// Return the size of the neighborhood.
+    size_type size() const { return neighborhood_.size(); }
+
+    /// Sort the neighborhood and sum the biases of duplicate variables.
+    void sort_and_sum() {
+        if (!std::is_sorted(begin(), end())) {
+            std::sort(begin(), end());
+        }
+
+        // now remove any duplicates, summing the biases of duplicates
+        size_type i = 0;
+        size_type j = 1;
+
+        // walk quickly through the neighborhood until we find a duplicate
+        while (j < neighborhood_.size() &&
+               neighborhood_[i].v != neighborhood_[j].v) {
+            ++i;
+            ++j;
+        }
+
+        // if we found one, move into de-duplication
+        while (j < neighborhood_.size()) {
+            if (neighborhood_[i].v == neighborhood_[j].v) {
+                neighborhood_[i].bias += neighborhood_[j].bias;
+                ++j;
+            } else {
+                ++i;
+                neighborhood_[i] = neighborhood_[j];
+                ++j;
+            }
+        }
+
+        // finally resize to contain only the unique values
+        // neighborhood_.resize(i + 1);
+        neighborhood_.erase(neighborhood_.begin() + i);
```

Comment:
I think it was throwing an error because of move not being properly implemented for `value_type`. I don't call `shrink_to_fit` because for most typical cases we're not saving a lot of space. So more performant to leave the capacity in there and avoid the copy.